### PR TITLE
Import MRSimulation files

### PR DESCRIPTION
**Pull request by philipturner**
MRSimulation is a new codec for efficiently replaying molecular simulation trajectories. Sometimes a simulation runs too slow to visualize in real-time, and if it takes several minutes to simulate, you want to save it to disk. Otherwise you must re-simulate every time the program starts up (<i>minutes</i> of startup latency). This was the original motivation behind molecular-renderer: decouple the compute cost of simulation from whether you can/can't visualize atoms moving over time.

This PR will let Linux/Windows users use [Molecular Renderer](https://github.com/philipturner/molecular-renderer) for design/simulation (until atomCAD gains such capabilities), and atomCAD for asynchronous visualization. I documented the codec and a benchmark suite at [MRSimulation.md](https://github.com/philipturner/molecular-renderer/blob/main/Documentation/MRSimulation.md), including tested decoders in Swift and Python.

However, I lack experience with the Rust language/ecosystem, and cannot finish the Rust decoder alone. This PR contains a ChatGPT-generated translation of the Swift version. I tried running it through the Rust web demo, and while I couldn’t interpret the errors, it failed to compile. Also unsure how this would integrate with atomCAD’s codebase (what namespace, class, or enum to encapsulate it under, etc.). Therefore it is a draft PR right now (allowing edits by maintainers).

<details>
<summary>Changes</summary>

- PDB files are <i>most often</i> static (no movement over time). This PR adds the ability to import MRSimulations, which <i>must</i> move over time.
  - Specifically, the plain-text variant: `.mrsim-txt`
  - Assumes no metadata is present (neither per-simulation nor per-frame metadata).
- This PR introduces no external library dependencies except `rayon` for multithreading
- May require modification to render pipeline, see Vsync/CVDisplayLink issues.
- Potential UI changes, described below.

</details>

<details>
<summary>Vsync/CVDisplayLink</summary>

- Needs some method in atomCAD to accurately track frame drops, to replay simulations without falling out of sync with IRL time.
  - <b>Solution 1:</b> Increment an integer counter by 1 each frame, so that dropped frames lag behind IRL time.
  - <b>Solution 2:</b> Calling a clock at program startup and doing `round_nearest_integer((current_time - startup_time) * refresh_rate)`. This creates rapid alternation between jumping 2 frames ahead, next the frame ID doesn’t increment at all, next it’s +1, 0, +1 ,+1, +2, +0, +1, +1, +1, etc…
- Hard-won lesson: no algorithm can fix both issues at once (tried some very complicated heuristics). Rather, it requires an OS-specific Vsync API handler, called once every display refresh cycle. The handler increments an atomic counter, and will be invoked on a background thread.
  - <b>How this fixes both problems.</b> During the run loop, you query the atomic counter from the main thread. Now time differences can be measured in integer numbers of frames (<b>Solution 2</b> is floating point), but unlike (<b>Solution 1</b>) it won’t lag with frame drops.
  - [CVDisplayLink](https://developer.apple.com/documentation/corevideo/cvdisplaylink-k0k) is unavailable from WebGPU, so atomCAD may need to choose one of the options above (stuttering or falling out of sync).
  - <b>Solution 3:</b> (never tried before) Interpolate frames based on the deviation between the current time and the nearest integer multiple of `1 / refresh_rate * n + startup_time`, where `n` is the discrete frame ID.

</details>

<details>
<summary>UI</summary>

- In my renderer, you hold the “P” key to start playing the simulation. The frame ID will increment every 1/120 second, until you release the key. “R” resets the trajectory to frame zero.
- Not sure how to implement this kind of playback in atomCAD’s UI, most likely through a menu bar item or 2D interface element.
- When atomCAD gets its own simulators, the most computationally intensive ones (ab initio quantum chemistry) may be too slow to run in real-time. Thus, a fallback option for simulating ahead of time can be a useful feature in general.

</details>

I hosted [18 pre-recorded trajectories](https://github.com/philipturner/mrsimulation-benchmarks) for testing this pull request against. They should all successfully replay in real-time before merging.

### CODE REVIEW

**Comment by crtag** [comment]
@philipturner I'm going to use your code sample in a bit different way for Rust, such as to produce both benchmark and a library atomCAD would use to render simulations in mrsim-text format. Will push updates into this PR.

**Comment by philipturner** [comment]
Sure, whatever path is most appropriate for Rust, I'm not very knowledgeable on that myself. An idea might be forking my branch of philipturner/atomCAD, then making a PR to my branch. I merge your commits, which become incorporated into the PR on atomCAD/atomCAD.

**Comment by crtag** [comment]
I can work on PR directly

**Comment by crtag** [comment]
@philipturner can you please upload something simpler (smaller file size) in mrsim-text format into the `assets` root folder?

**Comment by philipturner** [comment]
The smallest example I have [here](https://github.com/philipturner/mrsimulation-benchmarks/releases) is 6 MB. But I could make an MRSim with 1-2 of those 522-atom rhombic dodecahedra rotating, lasting only a second @ 120 fps. That should be under 1 MB.

**Comment by philipturner** [comment]
Trimmed my smallest simulation down to ~1 second (128 frames), which is 1 MB compressed. I can make a new crystolecule machine from scratch with <500 atoms suitable for testing.

[StrainedShellBearing-15ps.mrsim-txt.zip](https://github.com/atomCAD/atomCAD/files/12615550/StrainedShellBearing-15ps.mrsim-txt.zip)

**Comment by crtag** [comment]
Looks good, the only purpose is to run code tests. I can take other samples from your code base to do proper decoding.

**Comment by crtag** [comment]
It seems to be nearly impossible to work on other forks if I already have an active one myself. So the solution to work on PRs for forks the way the atomCAD main repository is set up now is to add me as a collaborator to your fork. I have some updates to this PR that I'm ready to push once you add me. The alternative for me is to fork your branch as you suggested, but I first need to delete my own fork. 
Basically, it means no one would be able to work on more than one PR at the time for atomCAD if there are no maintainers, or each PR owner would need to add contributors.

**Comment by philipturner** [comment]
I added you as a contributor to my fork.

**Comment by philipturner** [comment]
I would remove sp3 bonds from the metadata deserializer. That was just a proof of concept; no standardized format for bond topology has been created yet.

**Comment by crtag** [comment]
This is still a WiP, and I need to look into how this can be generalised with PDB and better abstracted. Performance seems to be OK-ish, with the release version it's slightly worse than Swift which could be due to the YAML parser overhead, but I'm keen to run it with all sample files. Anyways, I don't think parsing performance is critical.

**Comment by crtag** [comment]
> I would remove sp3 bonds from the metadata deserializer. That was just a proof of concept; no standardized format for bond topology has been created yet.

Not critical. I do want to understand the rationale behind clustering though.

**Comment by philipturner** [comment]
To get reasonable performance, you need to copy the text UTF-8 to a raw C buffer and operate on the characters as 8-bit integers. If you could report benchmarks of the Swift release vs your Rust parser, once you've verified the Rust code produces correct results. Use Markdown formatting for the table in GitHub comments:

```
| Time to Decode | Atoms | Unzipped Text Size      | Swift (Release) | Rust |
| ------------------------------ | ------ | ------ | ------ | ----- |
| Vdw Oscillator (Prototype 6)   | 10,000 | 58 MB  | ??? s  | ??? s |
| Vdw Oscillator (Final)         | 37,000 | 820 MB | ??? s  | ??? min |
| Strained Shell Bearing (15 ps) | 2,500  | 19 MB  | ??? s  | ??? s |
| Strained Shell Bearing (5 ns)  | 2,500  | 67 MB  | ??? s  | ??? s |
| Rhombic Dodecahedra (100 m/s)  | 34,000 | 285 MB | ??? s  | ??? min |
| Rhombic Dodecahedra (6400 m/s) | 34,000 | 316 MB | ??? s  | ??? min |
```

**Comment by philipturner** [comment]
> Not critical. I do want to understand the rationale behind clustering though.

MRSimulation uses delta compression between adjacent frames. The ideas is, differences in position store less information than absolute positions. If atoms are stationary, that might even create a string of zeroes. That would maximize the chance a ZIP compressor would compress them with run-length encoding.

After around 99% of the frames are stored as deltas, there's diminishing returns on compression ratio. Meanwhile, the overhead to access any single frame skyrockets (decode 128 just to access 1). I originally wanted this to be used in real-time with sparse decoding. Load (part of) the LZBITMAP-compressed simulation in RAM, decode a fraction of it at any one time. That can handle larger simulations than the current approach (decode everything at once upon startup).

Here's some research I did while creating the binary version of the codec: [Compression Ratios.pdf](https://github.com/atomCAD/atomCAD/files/12619337/Compression.Ratios.pdf)

**Comment by philipturner** [comment]
When parsing atom positions, I recommend casting to `f32` instead of `f64`. That is the default format for everything over at my code base, to make the code GPU compatible. For example, the binary MRSim decoder uses Float32 for GPU-accelerated decoding.

**Comment by crtag** [comment]
**Benchmark results**
Windows 10 Pro
AMD Ryzen 9 3900X 12-Core Processor, 3801 Mhz, 12 Core(s), 24 Logical Processor(s)
RAM 64.0 GB

| Time to Decode | Atoms | Unzipped Text Size      | Swift (Release) | Rust |
| ------------------------------ | ------ | ------ | ------ | ----- |
| Vdw Oscillator (Prototype 6)   | 10,000 | 58 MB  | ??? s  | 1.30 s |
| Vdw Oscillator (Final)         | 37,000 | 820 MB | ??? s  | 19.15 s |
| Strained Shell Bearing (15 ps) | 2,500  | 19 MB  | ??? s  | 0.47 s |
| Strained Shell Bearing (5 ns)  | 2,500  | 67 MB  | ??? s  | 1.55 s |
| Rhombic Dodecahedra (100 m/s)  | 34,000 | 285 MB | ??? s  | 6.97 s |
| Rhombic Dodecahedra (6400 m/s) | 34,000 | 316 MB | ??? s  | 7.05 s |

**Comment by philipturner** [comment]
Does the Rust version provide a section-by-section breakdown, profiling the latency of each step of decoding? That would be more directly comparable to Swift, provided you insert checkpoints at the same locations.

**Comment by crtag** [comment]
> Does the Rust version provide a section-by-section breakdown, profiling the latency of each step of decoding? That would be more directly comparable to Swift, provided you insert checkpoints at the same locations.

I'm going to try a couple of things to address both the performance and the latency profiling as you suggested. I'm not convinced it's essential at this stage, but worth investing in so we don't have to come back to this. More pressuring things is to get it married to atomCAD similarly to PDB format, i.e. get the elements indexed with the same enum, and, of course, deal with the animation "play" function. I suppose that's another large chunk of work.

There are also tests failing and I need to have a look at why this is happening, may need to ask in the Discord channel for assistance from Rust gurus.

**Comment by crtag** [comment]
I've added multi-threading and it did improve the performance a lot. Updated benchmarking as follows. Rayon picked 24 threads automatically in all cases.

**Benchmark results**
Windows 10 Pro
AMD Ryzen 9 3900X 12-Core Processor, 3801 Mhz, 12 Core(s), 24 Logical Processor(s)
RAM 64.0 GB

| Time to Decode | Atoms | Unzipped Text Size      | Swift (Release) | Rust |
| ------------------------------ | ------ | ------ | ------ | ----- |
| Vdw Oscillator (Prototype 6)   | 10,000 | 58 MB  | ??? s  | 0.37 s |
| Vdw Oscillator (Final)         | 37,000 | 820 MB | ??? s  | 2.5 s |
| Strained Shell Bearing (15 ps) | 2,500  | 19 MB  | ??? s  | 0.1 s |
| Strained Shell Bearing (5 ns)  | 2,500  | 67 MB  | ??? s  | 0.21 s |
| Rhombic Dodecahedra (100 m/s)  | 34,000 | 285 MB | ??? s  | 1.5 s |
| Rhombic Dodecahedra (6400 m/s) | 34,000 | 316 MB | ??? s  | 1.43 s |

**Comment by crtag** [comment]
> Does the Rust version provide a section-by-section breakdown, profiling the latency of each step of decoding? That would be more directly comparable to Swift, provided you insert checkpoints at the same locations.

I've replicated Swift reporting, so it splits parsing sections. Can be more granular if required.

**Comment by crtag** [comment]
@philipturner I've realized that Swift builds are single-threaded in Windows, at least the way I'm building it, that is either `swiftc -D RELEASE -Ounchecked <script-file-path>/NAME.swift` or `swiftc <script-file-path>/NAME.swift`. I need your expertise on this matter so I can provide the other side of the benchmark. Also, would be nice to get it going properly on Windows anyways as you aim for the cross-platform solution.

**Comment by philipturner** [comment]
Does this line return 1 in the Swift REPL? Try forcing it to 12 or 24 in the Swift script, by overwriting the line with an assignment of the variable to 12 or 24.

https://github.com/philipturner/molecular-renderer/blob/81cadd6e2c8f9cc321c79959f79f35c626f863cb/Documentation/Scripts/MRSimulationDecoder.swift#L374

**Comment by crtag** [comment]
> Does this line return 1 in the Swift REPL? Try forcing it to 12 or 24 in the Swift script, by overwriting the line with an assignment of the variable to 12 or 24.
> 
> https://github.com/philipturner/molecular-renderer/blob/81cadd6e2c8f9cc321c79959f79f35c626f863cb/Documentation/Scripts/MRSimulationDecoder.swift#L374

Yes, it picks 1. Overriding it with 24 gives better results, but still doesn't feel right.
Using **VdwOscillator-Final.mrsim-txt**
Built `swiftc  MRSimulationDecoder.swift`:
Loaded file in: 1.2 s
Preprocessed text in: 50.4 s
Parsed header in: 229.0 ms
Parsed clusters in: 7.2 s
Total decoding time: 59.1 s

Built `swiftc -D RELEASE  MRSimulationDecoder.swift`
Loaded file in: 1.5 s
Preprocessed text in: 7.6 s
Parsed header in: 252.0 ms
Parsed clusters in: 1.5 min
Total decoding time: 1.7 min

**Comment by philipturner** [comment]
`Preprocessed text in: 50.4 s`

This is the issue. There's a bottleneck in the stage of separating the text into separate strings based on newlines.

**Comment by crtag** [comment]
> `Preprocessed text in: 50.4 s`
> 
> This is the issue. There's a bottleneck in the stage of separating the text into separate strings based on newlines.

What about parse time being over a minute in RELEASE?

**Comment by philipturner** [comment]
My bad - I thought the 59 second one was release. The release mode parse time is worse than the debug mode parse time. That's not supposed to happen. Try swapping it out with the algorithm used for debug mode.

**Comment by philipturner** [comment]
`swiftc -D RELEASE  MRSimulationDecoder.swift`

It's not in release mode. You just ran debug mode with the release-mode optimizations that make debug mode significantly slower. Use the exact instructions posted on the README, which builds a `.exe` using `-Ounchecked` and then invoked the `.exe`.

**Comment by crtag** [comment]
how does the timing look like on Apple release vs debug? I'll play around with it later, my primary goal was to get frames/positions to validate the Rust version.

**Comment by philipturner** [comment]
I have a 10-core CPU with 3.2 GHz clock speed, similar performance to yours. Benchmarks are [here](https://github.com/philipturner/molecular-renderer/blob/main/Documentation/MRSimulation.md#benchmarks).

**Comment by crtag** [comment]
@philipturner updated with the correctly built swift 
**Benchmark results**
Windows 10 Pro
AMD Ryzen 9 3900X 12-Core Processor, 3801 Mhz, 12 Core(s), 24 Logical Processor(s)
RAM 64.0 GB
Both Swift and Rust are at 24 threads.

| Time to Decode | Atoms | Unzipped Text Size      | Swift (Release) | Rust |
| ------------------------------ | ------ | ------ | ------ | ----- |
| Vdw Oscillator (Prototype 6)   | 10,000 | 58 MB  | 1.2 s  | 0.37 s |
| Vdw Oscillator (Final)         | 37,000 | 820 MB | 9.4 s  | 2.5 s |
| Strained Shell Bearing (15 ps) | 2,500  | 19 MB  | 0.31 s  | 0.1 s |
| Strained Shell Bearing (5 ns)  | 2,500  | 67 MB  | 0.47 s  | 0.21 s |
| Rhombic Dodecahedra (100 m/s)  | 34,000 | 285 MB | 4.3 s  | 1.5 s |
| Rhombic Dodecahedra (6400 m/s) | 34,000 | 316 MB | 4.5 s  | 1.43 s |

**Comment by crtag** [comment]
@philipturner I plan to sort out GitHub CI checks issues.
Next steps up for discussion:

1. For any generic "play" mechanism I'd assume the rendering engine would rather use actual positions than deltas. So, the extra step after the parser is to turn its outcome into a continuous vector hash of atoms positions. This can be abstracted for atomCAD. This will wrap up this particular pull request. @maaku Thoughts?
2. A new, simulation "play" atomCAD development, I will update the [issue](https://github.com/atomCAD/atomCAD/issues/35)
3. A single-frame HDL results ->atomCAD functionality, potentially may come before #2 above.

**Comment by philipturner** [comment]
> philipturner updated with the correctly built swift
> Benchmark results
> Windows 10 Pro
> AMD Ryzen 9 3900X 12-Core Processor, 3801 Mhz, 12 Core(s), 24 Logical Processor(s)
> RAM 64.0 GB
> Both Swift and Rust are at 24 threads.

There seems to be something wrong with the Swift toolchain on Windows. I'll definitely take this into consider when making Windows builds of my framework. In general, anything super computationally intensive, you run on the GPU. CPU is for code you don't have much motivation to run fast, or have other constraints that make GPU not an appropriate solution. Or if you've been paywalled by Nvidia's CUDA and didn't find a way to bypass that paywall for your particular use case.

The Rust results line up very closely with the Swift release mode on macOS, so you're doing it right with Rust. No need to spend more time on the Swift script.

| Time to Decode | Atoms | Unzipped Text Size      | Swift, M1 Max | Rust, Ryzen 9 3900X |
| ------------------------------ | ------ | ------ | ------ | ----- |
| Vdw Oscillator (Prototype 6)   | 10,000 | 58 MB  | 0.3 s  | 0.37 s |
| Vdw Oscillator (Final)         | 37,000 | 820 MB | 3.6 s  | 2.5 s |
| Strained Shell Bearing (15 ps) | 2,500  | 19 MB  | 0.1 s  | 0.1 s |
| Strained Shell Bearing (5 ns)  | 2,500  | 67 MB  | 0.3 s  | 0.21 s |
| Rhombic Dodecahedra (100 m/s)  | 34,000 | 285 MB | 1.4 s  | 1.5 s |
| Rhombic Dodecahedra (6400 m/s) | 34,000 | 316 MB | 1.5 s  | 1.43 s |

**Comment by philipturner** [comment]
> A single-frame HDL results ->atomCAD functionality, potentially may come before the above point.

I would like that capability, but it doesn't seem atomCAD should have a dependency on the Swift Standard Library and toolchain. Mark suggested the idea of it being something of a plugin, like Eric's ideas for MSEP. Perhaps an out-of-tree fork to invoke the Swift libraries directly. I can improve capabilities of the JIT compiler so the DSL can be loaded in an atomCAD UI textbox.

**Comment by philipturner** [comment]
Okay, if this is working, we can start testing [mrsimulation-benchmarks](https://github.com/philipturner/mrsimulation-benchmarks).

Also, I did some order-of-magnitude calculations about the fastest way to offload trajectories from a Linux box with an 7900 XTX. Using a 400 MB/s USB 3 thumb drive (128 GB) with no data compression, just raw MRSimulation binary. This codec variant shouldn't be too hard to implement, and when I'm ready, I will make a Swift reference that can be translated to Rust.

https://github.com/philipturner/molecular-renderer/commit/5484d409b2c1eddb94b15aff1c1ff0bf95570fb9

**Comment by crtag** [comment]
@maaku ?

**Comment by maaku** [comment]
This was discussed on Discord, but for the sake of whoever is reading this in the future, this pull request was automatically closed in response to an accidental force-push to the `main` branch. I think we hit a GitHub bug, because even after restoring the branch to its original state, this PR cannot be reopened. A new pull request has already been created at #40
